src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+    auto offers = loadOffersByAccountAndAsset(account, asset);
+    for (auto const& offer : offers)
+    {
+        res[LedgerEntryKey(offer)] = offer;
+    }
+    return res;
+}
+
+LedgerHeader const&
+LedgerStateRoot::getHeader() const
+{
+    return mImpl->getHeader();
+}
+
+LedgerHeader const&
+LedgerStateRoot::Impl::getHeader() const
+{
+    return mHeader;
+}
+
+std::vector<InflationWinner>
+LedgerStateRoot::getInflationWinners(size_t maxWinners, int64_t minVotes)
+{
+    return mImpl->getInflationWinners(maxWinners, minVotes);
+}
+
+std::vector<InflationWinner>
+LedgerStateRoot::Impl::getInflationWinners(size_t maxWinners, int64_t minVotes)
+{
+    return loadInflationWinners(maxWinners, minVotes);
+}
+
+AbstractLedgerStateParent::EntryVersion
+LedgerStateRoot::getNewestVersion(LedgerKey const& key) const
+{
+    return mImpl->getNewestVersion(*this, key);
+}
+
+AbstractLedgerStateParent::EntryVersion
+LedgerStateRoot::Impl::getNewestVersion(LedgerStateRoot const& self,
+                                        LedgerKey const& key) const
+{
+    std::shared_ptr<LedgerEntry const> entry;
+    auto cacheKey = binToHex(xdr::xdr_to_opaque(key));
+    if (mCache->exists(cacheKey))
+    {
+        auto ptr = mCache->get(cacheKey);
+        entry = ptr ? std::make_shared<LedgerEntry const>(*ptr.get()) : nullptr;
+    }
+    else
+    {
+        switch (key.type())
+        {
+        case ACCOUNT:
+            entry = loadAccount(key);
+            break;
+        case DATA:
+            entry = loadData(key);
+            break;
+        case OFFER:
+            entry = loadOffer(key);
+            break;
+        case TRUSTLINE:
+            entry = loadTrustLine(key);
+            break;
+        default:
+            throw std::runtime_error("Unknown key type");
+        }
+        mCache->put(cacheKey, entry);
+    }
+
+    if (entry)
+    {
+        return {&self, entry};
+    }
+    return {nullptr, nullptr};
+}
+
+void
+LedgerStateRoot::rollbackChild(Identifier id)
+{
+    mImpl->rollbackChild();
+}
+
+void
+LedgerStateRoot::Impl::rollbackChild()
+{
+    mTransaction->rollback();
+    mTransaction.reset();
+    mChild = nullptr;
+}
+
+void
+LedgerStateRoot::Impl::storeAccount(
+    LedgerStateRoot const& self, LedgerKey const& key,
+    std::shared_ptr<LedgerEntry const> const& entry)
+{
+    auto const previous = getNewestVersion(self, key).entry;
+    if (entry)
+    {
+        insertOrUpdateAccount(*entry, !previous);
+        storeSigners(*entry, previous);
+    }
+    else if (previous) // Deleted
+    {
+        deleteAccount(key);
+    }
+}
+
+void
+LedgerStateRoot::Impl::storeData(
+    LedgerStateRoot const& self, LedgerKey const& key,
+    std::shared_ptr<LedgerEntry const> const& entry)
+{
+    auto const previous = getNewestVersion(self, key).entry;
+    if (entry)
+    {
+        insertOrUpdateData(*entry, !previous);
+    }
+    else if (previous) // Deleted
+    {
+        deleteData(key);
```

Comment:
why do we need to use `previous` when we delete? We could avoid loading the old value